api/src/main/java/org/apache/iceberg/types/Types.java

```diff
@@ -481,8 +534,19 @@ private NestedField(
       this.name = name;
       this.type = type;
       this.doc = doc;
-      this.initialDefault = initialDefault;
-      this.writeDefault = writeDefault;
+      this.initialDefault = castDefault(initialDefault, type);
+      this.writeDefault = castDefault(writeDefault, type);
+    }
+
+    private static Object castDefault(Object defaultValue, Type type) {
+      if (type.isNestedType() && defaultValue != null) {
+        throw new IllegalArgumentException(
+            String.format("Invalid default value for %s: %s (must be null)", type, defaultValue));
+      } else if (defaultValue != null) {
+        return Expressions.lit(defaultValue).to(type).value();
```

Comment:
In the description you mentioned that literals help with validations. Cannot we validated without bringing in literals?

Reply:
This should support the same values and coercion that we have implemented for literals, so I think this is the right way forward.